Mathlib/NumberTheory/RatFunc/Ostrowski.lean

```diff
@@ -0,0 +1,341 @@
+    (hab : v ↑(a * b) < 1 ∧ a ≠ 0 ∧ b ≠ 0) (hπᵥ : (WellFounded.min degree_lt_wf _ hne) = a * b)
+    (hb : ¬IsUnit b) : 1 ≤ v ↑a := by
+  set πᵥ := (WellFounded.min degree_lt_wf _ hne)
+  have hda : a.degree < πᵥ.degree := by
+    obtain hbpos := degree_pos_of_ne_zero_of_nonunit hab.2.2 hb
+    simp_rw [hπᵥ, degree_mul, degree_eq_natDegree hab.2.1, degree_eq_natDegree hab.2.2] at hbpos ⊢
+    norm_cast
+    simpa using hbpos
+  obtain hlea := imp_not_comm.mp (WellFounded.not_lt_min degree_lt_wf _ hne) hda
+  simp_all only [ne_eq, mem_setOf_eq, not_and, not_not, degree_mul, imp_false, not_lt]
+
+private theorem irreducible_min_polynomial_valuation_lt_one_and_ne_zero
+    (hne : {p : Fq[X] | v p < 1 ∧ p ≠ 0}.Nonempty)
+    (hle : v RatFunc.X ≤ 1) (h : ∀ a : Fq, a ≠ 0 → v (algebraMap Fq (RatFunc Fq) a) = 1) :
+    Irreducible (WellFounded.min degree_lt_wf {p : Fq[X] | v p < 1 ∧ p ≠ 0} hne) := by
+  set πᵥ := (WellFounded.min degree_lt_wf _ hne)
+  have hπᵥ : v πᵥ < 1 ∧ πᵥ ≠ 0 := WellFounded.min_mem degree_lt_wf _ hne
+  rw [irreducible_iff]
+  constructor
+  · simp only [Polynomial.isUnit_iff, isUnit_iff_ne_zero, ne_eq, not_exists, not_and]
+    intro a ha0 ha
+    absurd hπᵥ.1
+    rw [RatFunc.coePolynomial, ← ha, algebraMap_C, ← RatFunc.algebraMap_eq_C, h a ha0]
+    exact lt_irrefl 1
+  · by_contra!
+    obtain ⟨a, b, hab, hua, hub⟩ := this
+    rw [hab, mul_ne_zero_iff] at hπᵥ
+    suffices v a = 1 ∧ v b = 1 by simp_all [RatFunc.coePolynomial]
+    refine ⟨antisymm (Polynomial.valuation_le_one_of_valuation_X_le_one _ h hle _)
+        (one_le_valuation_factor hne hπᵥ hab hub),
+      antisymm (Polynomial.valuation_le_one_of_valuation_X_le_one _ h hle _) ?_⟩
+    simp only [mul_comm a b, And.comm (a := a ≠ 0)] at hπᵥ hab
+    exact one_le_valuation_factor hne hπᵥ hab hua
+
+section valuation_X_le_one
+
+variable (hle : v RatFunc.X ≤ 1) (h : ∀ (a : Fq), a ≠ 0 → v ((algebraMap Fq (RatFunc Fq)) a) = 1)
+   [v.IsNontrivial]
+
+/-- A uniformizing element for the valuation `v`, as a polynomial in `Fq[X]`. -/
+abbrev uniformizingPolynomial : Fq[X] :=
+  WellFounded.min degree_lt_wf _ (setOf_polynomial_valuation_lt_one_and_ne_zero_nonempty h hle)
+
+/-- A uniformizing element for the valuation `v`, as a polynomial in `Fq[X]`. -/
+local notation "πᵥ" => uniformizingPolynomial hle h
+
+lemma uniformizingPolynomial_ne_zero : ¬ πᵥ = 0 := by
+  obtain hπᵥ := WellFounded.min_mem degree_lt_wf _
+    ((setOf_polynomial_valuation_lt_one_and_ne_zero_nonempty h hle))
+  simp_all [uniformizingPolynomial]
+
+open Ideal in
+/-- The maximal ideal of `Fq[X]` generated by the `uniformizingPolynomial` for `v`. -/
+def valuationIdeal : HeightOneSpectrum Fq[X] where
+  asIdeal := Submodule.span Fq[X] {πᵥ}
+  isPrime := IsMaximal.isPrime (PrincipalIdealRing.isMaximal_of_irreducible
+    (irreducible_min_polynomial_valuation_lt_one_and_ne_zero
+      (setOf_polynomial_valuation_lt_one_and_ne_zero_nonempty h hle) hle h))
```

Comment:
Maybe make the ordering of `h` and `hle` consistent?
(But this may become a moot point if you decide to turn `h` into a type class.)